### PR TITLE
CLI Check

### LINKED ISSUE #1183: [Feature] Fission CLI check
Inspired by Linkerd CLI, I propose we add a check command to Fission CLI which can act in two scenarios:

1) Before installing Fission into the cluster - the user should be able to do a "pre-check" and see what exists or does not in the cluster for Fission to work properly. Potentially it can also emit YAMLs and install from CLI itself

```
$ fission check --pre
kubernetes
--------------
√ kubernetes version is compatible

storage
------------------
√ cluster can provision a disk

```

2) After Fission is installed in the cluster - a check everything is healthy!

```
$ fission check
fission
--------------
√ all fission controllers running fine

storage
------------------
√ cluster can provision a disk

fission-version
---------------
√ can determine the latest version
√ cli is up-to-date
```

### PR DESCRIPTION
**Pull request by shubham4443**
## Description

Added a new command which checks whether fission can run inside the cluster properly `fission check --pre`. Additionally, it also checks whether everything is healthy after installing fission `fission check`

### Sample outputs
- `fission check --pre`
image

- `fission check`
image
I believe because I was running in dev mode, my current fission version is showing 0.0.0

- Some failed cases -
image
image

## Which issue(s) this PR fixes:

Fixes #1183 

## Checklist:

- [ ] I ran tests as well as code linting locally to verify my changes. 
- [x] I have done manual verification of my changes, changes working as expected.
- [ ] I have added new tests to cover my changes.
- [x] My changes follow contributing guidelines of Fission.
- [ ] I have signed all of my commits.

### CODE REVIEW

File: cmd/fission-cli/app/app.go [hunk 1]
```diff
@@ -21,6 +22,7 @@ import (
 	"github.com/fission/fission/pkg/fission-cli/cmd/timetrigger"
 	"github.com/fission/fission/pkg/fission-cli/cmd/version"
 	"github.com/fission/fission/pkg/fission-cli/console"
+
```
Comment (sanketsudake) [inline cmd/fission-cli/app/app.go:25]: Please remove blank line.

File: cmd/fission-cli/app/app.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/fission/fission/pkg/fission-cli/cliwrapper/driver/cobra/helptemplate"
 	"github.com/fission/fission/pkg/fission-cli/cmd"
 	"github.com/fission/fission/pkg/fission-cli/cmd/canaryconfig"
+	"github.com/fission/fission/pkg/fission-cli/cmd/check"
```
Comment (sanketsudake) [inline cmd/fission-cli/app/app.go:13]: Looks like we are missing copyright in this file. Can you please add it ?
Reference: https://github.com/fission/fission/blob/3ac188124e348c2463a9e8b46bae9ab8967ce255/pkg/utils/gitrepo/gitrepo.go#L1

File: pkg/fission-cli/cmd/check/check.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package check
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/check/check.go:1]: Looks like we are missing copyright in this file. Can you please add it ?
Please do same for rest of new files.
Reference:

https://github.com/fission/fission/blob/3ac188124e348c2463a9e8b46bae9ab8967ce255/pkg/utils/gitrepo/gitrepo.go#L1

File: pkg/fission-cli/cmd/check/command.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package check
+
+import (
+	wrapper "github.com/fission/fission/pkg/fission-cli/cliwrapper/driver/cobra"
```
Comment (sanketsudake) [inline pkg/fission-cli/cmd/check/command.go:4]: Please fix this according to goimports standard.

File: pkg/healthcheck/healthcheck.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/fatih/color"
+	"github.com/fission/fission/pkg/controller/client"
+	"github.com/fission/fission/pkg/fission-cli/util"
```
Comment (sanketsudake) [inline pkg/healthcheck/healthcheck.go:11]: Please fix goimports

**Review by sanketsudake — CHANGES_REQUESTED** [review]

File: pkg/healthcheck/healthcheck.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+	version, err := hc.kubeAPI.ServerVersion()
+	if err != nil {
+		return err
+	}
+
+	major, _ := strconv.Atoi(version.Major)
+	minor, _ := strconv.Atoi(version.Minor)
+	apiVersion := [3]int{major, minor, 0}
+	minAPIVersion := [3]int{1, 19, 0}
```
Comment (shubham4443) [inline pkg/healthcheck/healthcheck.go:94]: Minimum Kubernetes version required is hardcoded. I want to confirm whether this should go inside `pkg/apis/core/v1/const.go

File: pkg/healthcheck/healthcheck.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+func (hc *HealthChecker) CheckFissionVersion() error {
+	ver := util.GetVersion(hc.FissionClient)
+
+	major, _ := strconv.Atoi(strings.Split(ver.Client["fission/core"].Version, ".")[0])
+	minor, _ := strconv.Atoi(strings.Split(ver.Client["fission/core"].Version, ".")[1])
+	fix, _ := strconv.Atoi(strings.Split(ver.Client["fission/core"].Version, ".")[2])
+
+	apiVersion := [3]int{major, minor, fix}
+	minAPIVersion := [3]int{1, 15, 1}
```
Comment (shubham4443) [inline pkg/healthcheck/healthcheck.go:131]: Latest fission released version is hardcoded. I want to confirm whether this should go inside `pkg/apis/core/v1/const.go`

File: pkg/healthcheck/healthcheck.go (hunk 6 above)
Comment (sanketsudake) [reply pkg/healthcheck/healthcheck.go:94]: I agree we should have constants for Kubernetes version. 
We can consider https://github.com/fission/fission/blob/master/pkg/fission-cli/util/constants.go

Comment (sanketsudake) [reply pkg/healthcheck/healthcheck.go:94]: This seems better place `pkg/apis/core/v1/const.go` Lets use `pkg/apis/core/v1/const.go`

File: pkg/healthcheck/healthcheck.go (hunk 7 above)
Comment (sanketsudake) [reply pkg/healthcheck/healthcheck.go:131]: We should check client vs server version here. Ideally we should require hardcoding.
client version is embedded in build info, where as server info is fetched from fission.
Client version should match with server version.

Comment (sanketsudake) [reply pkg/healthcheck/healthcheck.go:131]: Please refere server and client versions from https://github.com/fission/fission/blob/3ac188124e348c2463a9e8b46bae9ab8967ce255/pkg/fission-cli/util/util.go#L172

**Review by sanketsudake — CHANGES_REQUESTED** [review]

**Review by sanketsudake — APPROVED** [review]

**Pull request merged**